src/app/AttributePathExpandIterator.cpp

```diff
@@ -0,0 +1,188 @@
+    mAttributeIndex = UINT16_MAX;
+
+    // Make the iterator ready to emit the first valid path in the list.
+    Next();
+}
+
+void AttributePathExpandIterator::PrepareEndpointIndexRange(const ClusterInfo & aClusterInfo)
+{
+    if (aClusterInfo.HasWildcardEndpointId())
+    {
+        mBeginEndpointIndex = 0;
+        mEndEndpointIndex   = emberAfEndpointCount();
+    }
+    else
+    {
+        mBeginEndpointIndex = emberAfIndexFromEndpoint(aClusterInfo.mEndpointId);
+        // If the given cluster id does not exist on the given endpoint, it will return uint16(0xFFFF), then endEndpointIndex
+        // will be 0, means we should iterate a null endpoint set (skip it).
+        mEndEndpointIndex = static_cast<uint16_t>(mBeginEndpointIndex + 1);
+    }
+}
+
+void AttributePathExpandIterator::PrepareClusterIndexRange(const ClusterInfo & aClusterInfo, EndpointId aEndpointId)
+{
+    if (aClusterInfo.HasWildcardClusterId())
+    {
+        mBeginClusterIndex = 0;
+        mEndClusterIndex   = emberAfClusterCount(aEndpointId, true /* server */);
+    }
+    else
+    {
+        mBeginClusterIndex = emberAfClusterIndex(aEndpointId, aClusterInfo.mClusterId, CLUSTER_MASK_SERVER);
+        // If the given cluster id does not exist on the given endpoint, it will return uint8(0xFF), then endClusterIndex
+        // will be 0, means we should iterate a null cluster set (skip it).
+        mEndClusterIndex = static_cast<uint8_t>(mBeginClusterIndex + 1);
+    }
+}
+
+void AttributePathExpandIterator::PrepareAttributeIndexRange(const ClusterInfo & aClusterInfo, EndpointId aEndpointId,
+                                                             ClusterId aClusterId)
+{
+    if (aClusterInfo.HasWildcardAttributeId())
+    {
+        mBeginAttributeIndex = 0;
+        mEndAttributeIndex   = emberAfGetServerAttributeCount(aEndpointId, aClusterId);
+    }
+    else
+    {
+        mBeginAttributeIndex = emberAfGetServerAttributeIndexByAttributeId(aEndpointId, aClusterId, aClusterInfo.mAttributeId);
+        // If the given attribute id does not exist on the given endpoint, it will return uint16(0xFFFF), then endAttributeIndex
+        // will be 0, means we should iterate a null attribute set (skip it).
+        mEndAttributeIndex = static_cast<uint16_t>(mBeginAttributeIndex + 1);
+    }
+}
+
+bool AttributePathExpandIterator::Next()
+{
+    for (; mpClusterInfo != nullptr; (mpClusterInfo = mpClusterInfo->mpNext, mEndpointIndex = UINT16_MAX))
+    {
+        // Special case: If this is a concrete path, we just return its value as-is.
+        if (!mpClusterInfo->HasWildcard())
+        {
+            mOutputPath.mEndpointId  = mpClusterInfo->mEndpointId;
+            mOutputPath.mClusterId   = mpClusterInfo->mClusterId;
+            mOutputPath.mAttributeId = mpClusterInfo->mAttributeId;
+
+            // Prepare for next iteration
+            (mpClusterInfo = mpClusterInfo->mpNext, mEndpointIndex = UINT16_MAX);
+            return true;
```

Comment:
Another option would be to make the non-wildcard case more similar to wildcard case, where we keep pointing to the ClusterInfo we returned the path from, but mark ourselves as done with it.  Something like this:
```suggestion
            if (mEndpointIndex == UINT16_MAX) 
            {
                mOutputPath.mEndpointId  = mpClusterInfo->mEndpointId;
                mOutputPath.mClusterId   = mpClusterInfo->mClusterId;
                mOutputPath.mAttributeId = mpClusterInfo->mAttributeId;

                // Prepare for next iteration, when we'll want to move to the next ClusterInfo.
                mEndpointIndex = mEndEndpointIndex = 0;
                return true;
            }
            
            continue;
```
Then we can directly use `mpClusterInfo` in `Valid()`.